synapse/appservice/scheduler.py

```diff
@@ -212,7 +227,40 @@ async def _send_request(self, service: ApplicationService) -> None:
                 ]
                 del all_to_device_messages[:MAX_TO_DEVICE_MESSAGES_PER_TRANSACTION]
 
-                if not events and not ephemeral and not to_device_messages_to_send:
+                # Consolidate any pending device list summaries into a single, up-to-date
+                # summary.
+                # Note: this code assumes that in a single DeviceLists, a user will
+                # never be in both "changed" and "left" sets.
+                device_list_summary = DeviceLists()
+                while self.queued_device_list_summaries.get(service.id, []):
```

Comment:
The reason to `pop` items out of `self.queued_device_list_summaries` is so that we end up with an empty list for `self.queued_device_list_summaries[service.id]`.

But do you mean that this would be simpler by changing to a `for` loop, and emptying `self.queued_device_list_summaries.get[service.id]` once we're done?

Reply:
Yeah, I was thinking more of "loop over it, then clear it" instead of mutating it each time. Not a huge deal, just a bit odd I think?